src/main/java/software/amazon/event/kafkaconnector/batch/BestEffortInOrderEventBridgeBatching.java

```diff
@@ -0,0 +1,90 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package software.amazon.event.kafkaconnector.batch;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collector;
+import java.util.stream.Stream;
+import software.amazon.awssdk.services.eventbridge.model.PutEventsRequestEntry;
+import software.amazon.event.kafkaconnector.util.MappedSinkRecord;
+
+public class BestEffortInOrderEventBridgeBatching implements EventBridgeBatchingStrategy {
+
+  private final Collector<
+          MappedSinkRecord<PutEventsRequestEntry>,
+          PutEventsRequestEntrySizeCollector,
+          Stream<List<MappedSinkRecord<PutEventsRequestEntry>>>>
+      collector =
+          Collector.of(
+              PutEventsRequestEntrySizeCollector::new,
+              PutEventsRequestEntrySizeCollector::accumulate,
+              (left, right) -> {
+                throw new IllegalStateException();
+              },
+              PutEventsRequestEntrySizeCollector::finish);
+
+  @Override
+  public Stream<List<MappedSinkRecord<PutEventsRequestEntry>>> apply(
+      Stream<MappedSinkRecord<PutEventsRequestEntry>> records) {
+    if (records.isParallel()) {
+      throw new IllegalArgumentException("Stream must not be parallel.");
+    }
+    return records.collect(collector);
+  }
+
+  static class PutEventsRequestEntrySizeCollector {
+
+    private static final int MAX_BATCH_SIZE = 256 * 1024;
+
+    private final List<List<MappedSinkRecord<PutEventsRequestEntry>>> result = new ArrayList<>();
+    private int actualBatchSize = 0;
+    private int index = 0;
+
+    PutEventsRequestEntrySizeCollector() {
+      result.add(new ArrayList<>());
+    }
+
+    void accumulate(final MappedSinkRecord<PutEventsRequestEntry> item) {
+
+      int itemSize = getSize(item.getValue());
+      if (actualBatchSize + itemSize > MAX_BATCH_SIZE) {
+        result.add(new ArrayList<>());
+        index++;
+        actualBatchSize = 0;
+      }
+
+      actualBatchSize += itemSize;
+      result.get(index).add(item);
+    }
+
+    Stream<List<MappedSinkRecord<PutEventsRequestEntry>>> finish() {
+      return ((result.size() == 1) && (actualBatchSize == 0)) ? Stream.empty() : result.stream();
+    }
+
+    // https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-putevent-size.html
+    static int getSize(PutEventsRequestEntry entry) {
+      int size = 0;
+      if (entry.time() != null) {
+        size += 14;
+      }
+      size += entry.source().getBytes(UTF_8).length;
+      size += entry.detailType().getBytes(UTF_8).length;
+      if (entry.detail() != null) {
+        size += entry.detail().getBytes(UTF_8).length;
+      }
+      if (entry.resources() != null) {
+        for (String resource : entry.resources()) {
+          if (resource != null) {
+            size += resource.getBytes(UTF_8).length;
+          }
+        }
+      }
```

Comment:
Note: this is missing in the docs (yet) but `PutEvents` allows passing a `TraceHeader` which we also need to check: https://docs.aws.amazon.com/eventbridge/latest/APIReference/API_PutEventsRequestEntry.html

Reply:
:astonished: 

I can add this to `getSize` calculation. Currently we do not set the trace header when mapping from `SinkRecord`.

Reply:
Good point! Reason I'm bringing this up is that once we have an agreed implemention here, we want to update our docs too :) Treat this as a kind of reference implementation then

Reply:
Okay, will implement it.